// react-simplemde-editor/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SimpleMdeReact = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _easymde = _interopRequireDefault(require("easymde"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var _id = 0;

var generateId = function generateId() {
  return "simplemde-editor-".concat(++_id);
};

var useHandleEditorInstanceLifecycle = function useHandleEditorInstanceLifecycle(_ref) {
  var options = _ref.options,
      id = _ref.id,
      currentValueRef = _ref.currentValueRef,
      textRef = _ref.textRef;

  var _useState = (0, _react.useState)(null),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      editor = _useState2[0],
      setEditor = _useState2[1];

  var imageUploadCallback = (0, _react.useCallback)(function (file, onSuccess, onError) {
    var imageUpload = options === null || options === void 0 ? void 0 : options.imageUploadFunction;

    if (imageUpload) {
      var _onSuccess = function _onSuccess(url) {
        onSuccess(url);
      };

      imageUpload(file, _onSuccess, onError);
    }
  }, [options === null || options === void 0 ? void 0 : options.imageUploadFunction]);
  var editorRef = (0, _react.useRef)(editor);
  editorRef.current = editor;
  (0, _react.useEffect)(function () {
    var editor;

    if (textRef) {
      var initialOptions = {
        element: textRef,
        initialValue: currentValueRef.current
      };
      var imageUploadFunction = (options === null || options === void 0 ? void 0 : options.imageUploadFunction) ? imageUploadCallback : undefined;
      editor = new _easymde.default(Object.assign({}, initialOptions, options, {
        imageUploadFunction: imageUploadFunction
      }));
      setEditor(editor);
    }

    return function () {
      var _editor, _editor2;

      (_editor = editor) === null || _editor === void 0 ? void 0 : _editor.toTextArea(); // @ts-expect-error

      (_editor2 = editor) === null || _editor2 === void 0 ? void 0 : _editor2.cleanup();
    };
  }, [textRef, currentValueRef, id, imageUploadCallback, options]);
  var codemirror = (0, _react.useMemo)(function () {
    return editor === null || editor === void 0 ? void 0 : editor.codemirror;
  }, [editor === null || editor === void 0 ? void 0 : editor.codemirror]);
  return {
    editor: editor,
    codemirror: codemirror
  };
};

var SimpleMdeReact = /*#__PURE__*/_react.default.forwardRef(function (props, _ref8) {
  var events = props.events,
      value = props.value,
      options = props.options,
      children = props.children,
      extraKeys = props.extraKeys,
      getLineAndCursor = props.getLineAndCursor,
      getMdeInstance = props.getMdeInstance,
      getCodemirrorInstance = props.getCodemirrorInstance,
      onChange = props.onChange,
      anId = props.id,
      rest = (0, _objectWithoutProperties2.default)(props, ["events", "value", "options", "children", "extraKeys", "getLineAndCursor", "getMdeInstance", "getCodemirrorInstance", "onChange", "id"]);
  var id = (0, _react.useMemo)(function () {
    return anId !== null && anId !== void 0 ? anId : generateId();
  }, [anId]);
  var elementWrapperRef = (0, _react.useRef)(null);
  var nonEventChangeRef = (0, _react.useRef)(true); // This is to not pass value as a dependency e.g. to keep event handlers referentially
  // stable and do not `off` and `on` on each value change
  // plus to avoid unnecessary EasyEde editor recreation on each value change while still, if it has to be remounted
  // due to options and other deps change, to preserve that last value and not the default one from the first render.

  var currentValueRef = (0, _react.useRef)(value);
  currentValueRef.current = value;

  var _useState3 = (0, _react.useState)(null),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      textRef = _useState4[0],
      setTextRef = _useState4[1];

  var _useHandleEditorInsta = useHandleEditorInstanceLifecycle({
    options: options,
    id: id,
    currentValueRef: currentValueRef,
    textRef: textRef
  }),
      editor = _useHandleEditorInsta.editor,
      codemirror = _useHandleEditorInsta.codemirror;

  (0, _react.useEffect)(function () {
    // If change comes from the event we don't need to update `SimpleMDE` value as it already has it
    // Otherwise we shall set it as it comes from `props` set from the outside. E.g. by some reset button and whatnot
    if (nonEventChangeRef.current) {
      editor === null || editor === void 0 ? void 0 : editor.value(value !== null && value !== void 0 ? value : "");
    }

    nonEventChangeRef.current = true;
  }, [editor, value]);
  var onCodemirrorChangeHandler = (0, _react.useCallback)(function () {
    if ((editor === null || editor === void 0 ? void 0 : editor.value()) !== currentValueRef.current) {
      var _editor$value;

      nonEventChangeRef.current = false;
      onChange === null || onChange === void 0 ? void 0 : onChange((_editor$value = editor === null || editor === void 0 ? void 0 : editor.value()) !== null && _editor$value !== void 0 ? _editor$value : "");
    }
  }, [editor, onChange]);
  (0, _react.useEffect)(function () {
    // For some reason it doesn't work out of the box, this makes sure it's working correctly
    if (options === null || options === void 0 ? void 0 : options.autofocus) {
      codemirror === null || codemirror === void 0 ? void 0 : codemirror.focus();
      codemirror === null || codemirror === void 0 ? void 0 : codemirror.setCursor(codemirror === null || codemirror === void 0 ? void 0 : codemirror.lineCount(), 0);
    }
  }, [codemirror, options === null || options === void 0 ? void 0 : options.autofocus]);
  var getCursorCallback = (0, _react.useCallback)(function () {
    // https://codemirror.net/doc/manual.html#api_selection
    codemirror && (getLineAndCursor === null || getLineAndCursor === void 0 ? void 0 : getLineAndCursor(codemirror.getDoc().getCursor()));
  }, [codemirror, getLineAndCursor]);
  (0, _react.useEffect)(function () {
    getCursorCallback();
  }, [getCursorCallback]);
  (0, _react.useEffect)(function () {
    editor && (getMdeInstance === null || getMdeInstance === void 0 ? void 0 : getMdeInstance(editor));
  }, [editor, getMdeInstance]);
  (0, _react.useEffect)(function () {
    codemirror && (getCodemirrorInstance === null || getCodemirrorInstance === void 0 ? void 0 : getCodemirrorInstance(codemirror));
  }, [codemirror, getCodemirrorInstance, getMdeInstance]);
  (0, _react.useEffect)(function () {
    // https://codemirror.net/doc/manual.html#option_extraKeys
    if (extraKeys && codemirror) {
      codemirror.setOption("extraKeys", Object.assign({}, extraKeys, codemirror.getOption("extraKeys")));
    }
  }, [codemirror, extraKeys]);
  (0, _react.useEffect)(function () {
    var _elementWrapperRef$cu;

    var toolbarNode = (_elementWrapperRef$cu = elementWrapperRef.current) === null || _elementWrapperRef$cu === void 0 ? void 0 : _elementWrapperRef$cu.getElementsByClassName("editor-toolbarNode")[0];

    var handler = function handler() {
      return codemirror && onCodemirrorChangeHandler();
    };

    toolbarNode === null || toolbarNode === void 0 ? void 0 : toolbarNode.addEventListener("click", handler);
    return function () {
      toolbarNode === null || toolbarNode === void 0 ? void 0 : toolbarNode.removeEventListener("click", handler);
    };
  }, [codemirror, onCodemirrorChangeHandler]);
  (0, _react.useEffect)(function () {
    codemirror === null || codemirror === void 0 ? void 0 : codemirror.on("change", onCodemirrorChangeHandler);
    codemirror === null || codemirror === void 0 ? void 0 : codemirror.on("cursorActivity", getCursorCallback);
    return function () {
      codemirror === null || codemirror === void 0 ? void 0 : codemirror.off("change", onCodemirrorChangeHandler);
      codemirror === null || codemirror === void 0 ? void 0 : codemirror.off("cursorActivity", getCursorCallback);
    };
  }, [codemirror, getCursorCallback, onCodemirrorChangeHandler]);
  var prevEvents = (0, _react.useRef)(events);
  (0, _react.useEffect)(function () {
    var isNotFirstEffectRun = events !== prevEvents.current;
    isNotFirstEffectRun && prevEvents.current && Object.entries(prevEvents.current).forEach(function (_ref2) {
      var _ref3 = (0, _slicedToArray2.default)(_ref2, 2),
          event = _ref3[0],
          handler = _ref3[1];

      handler && (codemirror === null || codemirror === void 0 ? void 0 : codemirror.off(event, handler));
    });
    events && Object.entries(events).forEach(function (_ref4) {
      var _ref5 = (0, _slicedToArray2.default)(_ref4, 2),
          event = _ref5[0],
          handler = _ref5[1];

      handler && (codemirror === null || codemirror === void 0 ? void 0 : codemirror.on(event, handler));
    });
    prevEvents.current = events;
    return function () {
      events && Object.entries(events).forEach(function (_ref6) {
        var _ref7 = (0, _slicedToArray2.default)(_ref6, 2),
            event = _ref7[0],
            handler = _ref7[1];

        handler && (codemirror === null || codemirror === void 0 ? void 0 : codemirror.off(event, handler));
      });
    };
  }, [codemirror, events]);
  return /*#__PURE__*/_react.default.createElement("div", Object.assign({
    id: "".concat(id, "-wrapper")
  }, rest, {
    ref: function ref(aRef) {
      if (typeof _ref8 === "function") {
        _ref8(aRef);
      } else if (_ref8) {
        _ref8.current = aRef;
      }

      elementWrapperRef.current = aRef;
    }
  }), /*#__PURE__*/_react.default.createElement("textarea", {
    id: id,
    ref: setTextRef,
    style: {
      display: "none"
    }
  }));
});

exports.SimpleMdeReact = SimpleMdeReact;
SimpleMdeReact.displayName = "SimpleMdeReact";
var _default = SimpleMdeReact;
exports.default = _default;
